source/grid/manifold.cc

```diff
@@ -662,6 +697,223 @@ FlatManifold<dim, spacedim>::get_tangent_vector (const Point<spacedim> &x1,
+{
+  const Tensor<1,2> tangent = face->vertex(1) - face->vertex(0);
+  for (unsigned int vertex=0; vertex<GeometryInfo<2>::vertices_per_face; ++vertex)
+    // compute normals from tangent
+    face_vertex_normals[vertex] = Point<2>(tangent[1],
+                                           -tangent[0]);
+}
+
+
+
+template <>
+void
+FlatManifold<2,3>::
+get_normals_at_vertices (const Triangulation<2,3>::face_iterator &face,
+                         Manifold<2,3>::FaceVertexNormals &face_vertex_normals) const
+{
+  Assert(false, ExcNotImplemented());
+}
+
+
+
+template <>
+void
+FlatManifold<3>::
+get_normals_at_vertices (const Triangulation<3>::face_iterator &face,
+                         Manifold<3,3>::FaceVertexNormals &face_vertex_normals) const
+{
+  const unsigned int vertices_per_face = GeometryInfo<3>::vertices_per_face;
+
+  static const unsigned int neighboring_vertices[4][2]=
+  { {1,2},{3,0},{0,3},{2,1}};
+  for (unsigned int vertex=0; vertex<vertices_per_face; ++vertex)
+    {
+      // first define the two tangent vectors at the vertex by using the
+      // two lines radiating away from this vertex
+      const Tensor<1,3> tangents[2]
+        = { face->vertex(neighboring_vertices[vertex][0])
+            - face->vertex(vertex),
+            face->vertex(neighboring_vertices[vertex][1])
+            - face->vertex(vertex)
+          };
+
+      // then compute the normal by taking the cross product. since the
+      // normal is not required to be normalized, no problem here
+      face_vertex_normals[vertex] = cross_product_3d(tangents[0], tangents[1]);
+    }
+}
+
+
+
+template <>
+Tensor<1,1>
+FlatManifold<1,1>::
+normal_vector (const Triangulation<1,1>::face_iterator &,
+               const Point<1> &) const
+{
+  Assert (false, ExcNotImplemented());
+  return Tensor<1,1>();
+}
+
+
+
+template <>
+Tensor<1,2>
+FlatManifold<1,2>::
+normal_vector (const Triangulation<1,2>::face_iterator &,
+               const Point<2> &) const
+{
+  Assert (false, ExcNotImplemented());
+  return Tensor<1,2>();
+}
+
+
+
+template <>
+Tensor<1,3>
+FlatManifold<1,3>::
+normal_vector (const Triangulation<1,3>::face_iterator &,
+               const Point<3> &) const
+{
+  Assert (false, ExcNotImplemented());
+  return Tensor<1,3>();
+}
+
+
+
+template <int dim, int spacedim>
+Tensor<1, spacedim>
+FlatManifold< dim, spacedim >::
+normal_vector (const typename Triangulation<dim, spacedim>::face_iterator &face,
+               const Point<spacedim>                                      &p) const
+{
+  // I don't think the implementation below will work when dim!=spacedim;
+  // in fact, I believe that we don't even have enough information here,
+  // because we would need to know not only about the tangent vectors
+  // of the face, but also of the cell, to compute the normal vector.
+  // Someone will have to think about this some more.
+  Assert (dim == spacedim, ExcNotImplemented());
```

Comment:
so given the specializations above, this can then only cover the <2,2> and <3,3> cases, right? Shouldn't the <2,2> case be easy because a face then is just a straight line whose normal vector we can easily compute? I think only the <3,3> case is difficult.

Reply:
Yes, thanks for the suggestion. I implemented the specialization.
This caused `grid/normal_vector_03_2d` to fail which turned out to be an incorrect equivalent of the 3d case.